File: feature/matrix/src/main/java/net/twisterrob/colorfilters/android/matrix/Component.java
package net.twisterrob.colorfilters.android.matrix;

import android.content.SharedPreferences;
import android.graphics.ColorMatrix;
import android.view.*;
import android.widget.*;

// TODO refactor to subfragments and/or custom View/ViewGroups
abstract class Component {
	interface RefreshListener {
		void refresh(boolean recombine);
	}

	protected static final int COMP_R = 0;
	protected static final int COMP_G = 1;
	protected static final int COMP_B = 2;
	protected static final int COMP_A = 3;

	protected final View view;
	private final RefreshListener listener;

	public Component(View view, RefreshListener listener) {
		this.view = view;
		this.listener = listener;
	}

	public abstract void setupUI();

	public void unWire() {
	}

	protected EditText et(int viewID) {
		return (EditText)view.findViewById(viewID);
	}

	protected TextView tv(int viewID) {
		return (TextView)view.findViewById(viewID);
	}

	protected View v(int viewID) {
		return view.findViewById(viewID);
	}

	protected ViewGroup vg(int viewID) {
		return (ViewGroup)view.findViewById(viewID);
	}

	protected SeekBar sb(int viewID) {
		return (SeekBar)view.findViewById(viewID);
	}

	public abstract void reset();

	public abstract void refreshModel();

	public abstract void combineInto(ColorMatrix colorMatrix);

	public abstract boolean appendTo(StringBuilder sb);

	protected static float get(SeekBar seekBar, float scale, float offset) {
		return (float)seekBar.getProgress() / (float)seekBar.getMax() * scale + offset;
	}

	protected static void set(SeekBar seekBar, float scale, float offset, float value) {
		seekBar.setProgress((int)((value - offset) / scale * seekBar.getMax()));
	}

	protected void dispatchRefresh(boolean recombine) {
		listener.refresh(recombine);
	}

	public abstract void saveToPreferences(SharedPreferences.Editor editor);

	public abstract void restoreFromPreferences(SharedPreferences prefs);
}

File: feature/matrix/src/main/kotlin/net/twisterrob/colorfilters/android/matrix/Component.kt
package net.twisterrob.colorfilters.android.matrix

import android.content.SharedPreferences
import android.graphics.ColorMatrix
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.SeekBar
import android.widget.TextView

// TODO refactor to subfragments and/or custom View/ViewGroups
internal abstract class Component(
	protected val view: View,
	private val listener: RefreshListener
) {

	internal interface RefreshListener {
		fun refresh(recombine: Boolean)
	}

	abstract fun setupUI()

	open fun unWire() {}

	protected fun et(viewID: Int): EditText = view.findViewById(viewID)
	protected fun tv(viewID: Int): TextView = view.findViewById(viewID)
	protected fun v(viewID: Int): View = view.findViewById(viewID)
	protected fun vg(viewID: Int): ViewGroup = view.findViewById(viewID)
	protected fun sb(viewID: Int): SeekBar = view.findViewById(viewID)

	abstract fun reset()

	abstract fun refreshModel()

	abstract fun combineInto(colorMatrix: ColorMatrix)

	abstract fun appendTo(sb: StringBuilder): Boolean

	protected fun dispatchRefresh(recombine: Boolean) {
		listener.refresh(recombine)
	}

	abstract fun saveToPreferences(editor: SharedPreferences.Editor)

	abstract fun restoreFromPreferences(prefs: SharedPreferences)

	companion object {

		@JvmStatic val COMP_R = 0
		@JvmStatic val COMP_G = 1
		@JvmStatic val COMP_B = 2
		@JvmStatic val COMP_A = 3

		@JvmStatic
		protected fun get(seekBar: SeekBar, scale: Float, offset: Float): Float =
			seekBar.progress.toFloat() / seekBar.max.toFloat() * scale + offset

		@JvmStatic
		protected fun set(seekBar: SeekBar, scale: Float, offset: Float, value: Float) {
			seekBar.progress = ((value - offset) / scale * seekBar.max).toInt()
		}
	}
}
